File: app/src/main/java/com/sjn/stamp/db/TotalSongHistory.java
package com.sjn.stamp.db;

import com.sjn.stamp.constant.RecordType;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

@Accessors(prefix = "m")
@AllArgsConstructor(suppressConstructorProperties = true)
@NoArgsConstructor
@Getter
@Setter
public class TotalSongHistory extends RealmObject {
    @PrimaryKey
    public long mId;
    private Song mSong;

    public int mPlayCount;
    public int mSkipCount;
    public int mCompleteCount;

    public void setPlayCountIfOver(int playCount) {
        if (playCount <= mPlayCount) {
            return;
        }
        mPlayCount = playCount;
    }

    public void setSkipCountIfOver(int skipCount) {
        if (skipCount <= mSkipCount) {
            return;
        }
        mSkipCount = skipCount;
    }

    public void setCompleteCountIfOver(int completeCount) {
        if (completeCount <= mCompleteCount) {
            return;
        }
        mCompleteCount = completeCount;
    }

    public void incrementPlayCount(int playCount) {
        mPlayCount += playCount;
    }

    public void incrementSkipCount(int skipCount) {
        mSkipCount += skipCount;
    }

    public void incrementCompleteCount(int completeCount) {
        mCompleteCount += completeCount;
    }

    public void parseSongQueue(Song song, RecordType recordType) {
        switch (recordType) {
            case PLAY:
                setValues(song, 1, 0, 0);
                break;
            case SKIP:
                setValues(song, 0, 1, 0);
                break;
            case COMPLETE:
                setValues(song, 0, 0, 1);
                break;
            default:
                setValues(song, 0, 0, 0);
                break;
        }
    }

    public void setValues(Song song, int playCount, int skipCount, int completeCount) {
        setSong(song);
        setPlayCount(playCount);
        setSkipCount(skipCount);
        setCompleteCount(completeCount);
    }
}

File: app/src/main/java/com/sjn/stamp/db/TotalSongHistory.kt
package com.sjn.stamp.db

import com.sjn.stamp.constant.RecordType
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey

open class TotalSongHistory(
        @PrimaryKey
        var id: Long = 0,
        var song: Song = Song(),
        var playCount: Int = 0,
        var skipCount: Int = 0,
        var completeCount: Int = 0
) : RealmObject() {

    fun updatePlayCountIfOver(playCount: Int) {
        if (playCount <= this.playCount) {
            return
        }
        this.playCount = playCount
    }

    fun updateSkipCountIfOver(skipCount: Int) {
        if (skipCount <= this.skipCount) {
            return
        }
        this.skipCount = skipCount
    }

    fun updateCompleteCountIfOver(completeCount: Int) {
        if (completeCount <= this.completeCount) {
            return
        }
        this.completeCount = completeCount
    }

    fun incrementPlayCount(playCount: Int) {
        this.playCount += playCount
    }

    fun incrementSkipCount(skipCount: Int) {
        this.skipCount += skipCount
    }

    fun incrementCompleteCount(completeCount: Int) {
        this.completeCount += completeCount
    }

    fun applySongQueue(song: Song, recordType: RecordType): TotalSongHistory {
        when (recordType) {
            RecordType.PLAY -> applyValues(song, 1, 0, 0)
            RecordType.SKIP -> applyValues(song, 0, 1, 0)
            RecordType.COMPLETE -> applyValues(song, 0, 0, 1)
            else -> applyValues(song, 0, 0, 0)
        }
        return this
    }

    private fun applyValues(song: Song, playCount: Int, skipCount: Int, completeCount: Int) {
        this.song = song
        this.playCount = playCount
        this.skipCount = skipCount
        this.completeCount = completeCount
    }
}
